Sources/Core/ThreadSafeQueue.swift

```diff
@@ -30,13 +30,13 @@ class ThreadSafeQueue<T> {
                 return
             }
 
+            var firstItem: T?
+            if !self.list.isEmpty {
+                firstItem = self.list.removeFirst()
+            }
+
             self.completionQueue.async {
-                if !self.list.isEmpty {
-                    completion(self.list.removeFirst())
-                }
-                else {
-                    completion(nil)
-                }
+                completion(firstItem)
```

Comment:
Can you explain this change, please?

Reply:
To ensure that ThreadSafeQueue is really safe, all operations like `enqueue`, `dequeue`, `isEmpty` and `peek` should be executed on the same serial queue which is in our case `utilityQueue`. 
Otherwise a race condition can occurs which happened in our case. 

In `ThreadSafeQueue` class `enqueue` and `isEmpty` methods are accessing our resource which is a `private var list = [T]()` property in `utilityQueue`, but `dequeue` and `peek`
are accessing to `list` property on different queue `completionQueue`. So it is possible that two separeted threads accessing our `list` property at the same time hence race condition.

To fix that I changed that all operations on `list` property are executed on `utilityQueue` and `completionQueue` only gets the result.

Reply:
Cool - I see that now, on line 28!  Thanks!